sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala

```diff
@@ -186,6 +186,10 @@ class StateSchemaCompatibilityChecker(
           check(existingStateSchema, newSchema, ignoreValueSchema)
         }
       }
+      // we want to write the new schema file on each new run for state schema v3
+      if (stateSchemaVersion == 3) {
+        createSchemaFile(newStateSchemaList, stateSchemaVersion)
```

Comment:
We only want to create a new state schema file if it has changed in a compatible way right ?

Reply:
Hopefully the check I added is enough? I didn't want to get into the semantics of evolution in this PR, which is why I didn't modify the `check` method or use that result anywhere here.